### PR TITLE
Arbitrum L1 to L2 messaging proposal

### LINKED ISSUE #13515: Add Proposal example for Arbitrum L1 to L2 Messaging (aka Retryable tickets)
Add an example of how to create a DAO proposal to execute arbitrum's L1 to L2 messaging also known as Retryable tickets to the ```governance/proposal/```

### PR DESCRIPTION
**Pull request by blahkheart**
# Description

This PR adds an example of how to create a DAO proposal for execution of Arbitrum's L1-to-L2 message passing system (aka "retryable tickets")

# Issues

Fixes #13515
Refs #13515

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: .clabot
Comment (clemsos) [inline .clabot:1]: This should be moved to another PR

File: governance/.env
Comment (clemsos) [inline governance/.env:1]: Please remove the .env files are we don't support these

File: governance/.env [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+export ARBISCAN_API_KEY=
+export CELO_API_KEY=
+export GNOSISSCAN_API_KEY=
+
+# This is a sample .env file for use in local development.
+
+# Duplicate this file as .env
+
+# Your Private key
+DEVNET_PRIVKEY="0x14e1a3103ce06d0348acf19df3df1965ee55f9cdb16708c0bb8cd3d843ab8805"
```
Comment (clemsos) [inline governance/.env:19]: You should never check a private key into a git repo...

File: governance/.env [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+# Hosted Aggregator Node (JSON-RPC Endpoint). This is Arbitrum Sepolia Testnet, can use any Arbitrum chain
+
+L2RPC="https://arbitrum-mainnet.infura.io/v3/988a82bfe0a540ad84aa1a5d534c2bc6"
```
Comment (clemsos) [inline governance/.env:23]: Please move these into the proposal file

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+  console.log(':::::::::L1ToL2MessageGasParams::::::::::')
+  console.log(
+    'GasParams::::gasLimit',
+    L1ToL2MessageGasParams.gasLimit.toNumber()
+  )
+  console.log(
+    'GasParams::::maxSubmissionCost',
+    L1ToL2MessageGasParams.maxSubmissionCost.toNumber()
+  )
+  console.log(
```
Comment (clemsos) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:188]: Could you please remove the console logs from the file ? Better to keep things clean

**Review by clemsos — CHANGES_REQUESTED** [review]
Thanks for the PR. A few comments added !

File: governance/.env (hunk 1 above)
Comment (blahkheart) [reply governance/.env:19]: Yeah totally agree with that, I didn't know .env wasn't included in the .gitignore file, especially because I saw a .env.copy file already present in the ```/governance/``` folder, which I copied and renamed to .env. For what it's worth it's a fresh wallet I created specifically for the tests.

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 3 above)
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js]: All done

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+const walletPrivateKey = 'Your_Private_key'
+const L1RPC = 'https://mainnet.infura.io/v3/<your_infura_api_key>'
+const L2RPC = 'https://arbitrum-mainnet.infura.io/v3/<your_infura_api_key>'
+const l1Provider = new ethers.JsonRpcProvider(L1RPC)
+const l2Provider = new ethers.JsonRpcProvider(L2RPC)
+// const l1Wallet = new ethers.Wallet(walletPrivateKey, l1Provider)
+const l2Wallet = new ethers.Wallet(walletPrivateKey, l2Provider)
+
+module.exports = async () => {
+  await arbLog('Cross-chain Proposer')
```
Comment (clemsos) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:127]: what is `arbLog` used for ?

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        ARBTokenAddressOnL2,
+        0,
+        L1ToL2MessageGasParams.maxSubmissionCost,
+        l2Wallet.address,
+        l2Wallet.address,
+        L1ToL2MessageGasParams.gasLimit,
+        gasPriceBid,
+        transfer_calldata,
+      ],
+      value: L1ToL2MessageGasParams.deposit.toNumber() * 10, // I Multiply by 10 to add extra in case gas changes due to proposal delay
```
Comment (clemsos) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:217]: Are you sure you need to pass the value again here? It is already declared above

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 4 above)
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:127]: It came with the arbitrum sdk, its more or less a branding/description of the script.

File: governance/.env.copy [hunk 6]
```diff
@@ -9,4 +9,19 @@ export BSCSCAN_API_KEY=
 export SNOWTRACE_API_KEY=
 export ARBISCAN_API_KEY=
 export CELO_API_KEY=
-export GNOSISSCAN_API_KEY=
\ No newline at end of file
+export GNOSISSCAN_API_KEY=
+
+# This is a sample .env file for use in local development.
```
Comment (julien51) [inline governance/.env.copy:14]: Please remeove these changes...

File: governance/package.json [hunk 7]
```diff
@@ -3,6 +3,8 @@
   "description": "Scripts for the management of the Unlock Protocol",
   "private": true,
   "dependencies": {
+    "@arbitrum/nitro-contracts": "v1.0.2",
+    "@arbitrum/sdk": "^v3.1.9",
```
Comment (julien51) [inline governance/package.json:7]: Please use fixed version numbers (no `v` or `^`)

File: governance/package.json [hunk 8]
```diff
@@ -21,8 +23,9 @@
     "@unlock-protocol/hardhat-helpers": "workspace:^",
     "@unlock-protocol/hardhat-plugin": "workspace:^",
     "@unlock-protocol/networks": "workspace:./packages/networks",
+    "arb-shared-dependencies": "^1.0.0",
```
Comment (julien51) [inline governance/package.json:26]: Same here!

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    ],
+    stateMutability: 'payable',
+    type: 'function',
+  },
+]
+
+/**
+ * Set up: instantiate L1 / L2 wallets connected to providers
+ */
+const walletPrivateKey = 'Your_Private_key'
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:118]: Please use env variables here

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+const L1RPC = 'https://mainnet.infura.io/v3/<your_infura_api_key>'
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:119]: Please use `https://rpc.unlock-protocol.com/1` so that this does not break in the future.

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 4 above)
Comment (julien51) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:127]: Can we just use a regular `console.log` please?

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+const l2Wallet = new ethers.Wallet(walletPrivateKey, l2Provider)
+
+module.exports = async () => {
+  await arbLog('Cross-chain Proposer')
+  addDefaultLocalNetwork()
+
+  const l2Network = await getL2Network(l2Provider)
+  const ethBridger = new EthBridger(l2Network)
+  const inboxAddress = ethBridger.l2Network.ethBridge.inbox
+  const ARBTokenAddressOnL2 = '0x912CE59144191C1204E64559FE8253a0e49E6548' // ARB TOKEN ADDRESS ON ARBITRUM ONE
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:133]: Can yoiu please use the address (or add it there) in the `package/networks` file?

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+  const grantsContractAddress = '0x00D5E0d31d37cc13C645D86410aB4cB7Cb428ccA'
+  const timelockL2Alias = '0x28ffDfB0A6e6E06E95B3A1f928Dc4024240bD87c' // Timelock Alias Address on L2
+  const L1TimelockContract = '0x17EEDFb0a6E6e06E95B3A1F928dc4024240BC76B' // Timelock Address mainnet
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:136]: We should probably move that to a config file in this folder.

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+   * (1) maxSubmissionCost: The maximum cost to be paid for submitting the transaction
+   * (2) gasLimit: The L2 gas limit
+   * (3) deposit: The total amount to deposit on L1 to cover L2 gas and L2 call value
+   */
+  const L1ToL2MessageGasParams = await l1ToL2MessageGasEstimate.estimateAll(
+    {
+      from: L1TimelockContract,
+      to: ARBTokenAddressOnL2,
+      l2CallValue: 0,
+      excessFeeRefundAddress: l2Wallet.address,
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:172]: Can you explain why `l2Wallet` would recive the excess here?

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+      callValueRefundAddress: l2Wallet.address,
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:173]: Same here?

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+    },
+    await getBaseFee(l1Provider),
+    l1Provider
+  )
+  const gasPriceBid = await l2Provider.getGasPrice()
+
+  const inbox_calldata = iface_inbox.encodeFunctionData(
+    'createRetryableTicket',
+    [
+      ARBTokenAddressOnL2,
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:184]: Please add comments in front of each of these params to provide their names

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+      L1ToL2MessageGasParams.maxSubmissionCost,
+      l2Wallet.address,
+      l2Wallet.address,
+      L1ToL2MessageGasParams.gasLimit,
+      gasPriceBid,
+      transfer_calldata,
+    ]
+  )
+
+  const proposalName = `# Test Transaction before 7k ARB Transfer To Fund Unlock Protocol’s Ecosystem via Grants Stack  This proposal requests to use 1 ARB from the tokens given to Unlock Protocol DAO by ArbitrumDAO to run a test transaction to de-risk the transfer of 7k ARB tokens to fund the retroQF round on Grants Stack.`
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:195]: Please format this correctly. 

This would be 

# Test Transaction before 7k ARB Transfer To Fund Unlock Protocol’s Ecosystem via Grants Stack This proposal requests to use 1 ARB from the tokens given to Unlock Protocol DAO by ArbitrumDAO to run a test transaction to de-risk the transfer of 7k ARB tokens to fund the retroQF round on Grants Stack.

Make sure there is a title, one (or more) distinct paragraphs, as well as links to provide clarity (links to the snapshot, to the UI on which people are voting... etc)

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 5 above)
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:217]: Yes, from my understanding of the existing format, and correct me if I'm wrong but this value property is responsible for passing the eth amount to a payable function, and in this case that's the eth deposit to be sent to arbitrum's inbox contract. 
If you were referring to this line ```const values = [L1ToL2MessageGasParams.deposit.toNumber() * 10]``` that was declared to show how the proposal can be created from an interface like etherscan or remix, but if its not necessary or confusing I can remove it as well.

**Review by julien51 — CHANGES_REQUESTED** [review]
Can you please use https://github.com/unlock-protocol/unlock/pull/13509 to run tests for your proposal, then import the values in Tenderly to check that it passes?

File: governance/.env.copy (hunk 6 above)
Comment (blahkheart) [reply governance/.env.copy:14]: Okay got it

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 9 above)
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:118]: @clemsos earlier informed me .env is not supported, how else would I use env variables?

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 14 above)
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:173]: Same reason as before

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 9 above)
Comment (julien51) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:118]: Process.env.PRIVATE_KEY for example

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 13 above)
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js]: Got it, it has now been changed to the DAO's address. The reason for having it that way in the first place is due to the nature of the transaction. The ```createRetryable``` function forces the caller to pay the value sent to the Delayed Inbox contract, Therefore ETH is paid by the person calling the ```execute()``` function and not the Timelock. So it was set up so whoever creates the proposal and provides the private key gets any excess ETH refund sent to that address. But that was flawed by the assumption that the proposer will always be the executor. Ideally, the DAO will fund/reimburse whoever executes the proposal so excess should return to the DAO, thank you for pointing that out.

File: governance/package.json [hunk 17]
```diff
@@ -21,8 +23,9 @@
     "@unlock-protocol/hardhat-helpers": "workspace:^",
     "@unlock-protocol/hardhat-plugin": "workspace:^",
     "@unlock-protocol/networks": "workspace:./packages/networks",
+    "arb-shared-dependencies": "1.0.0",
     "eslint": "8.54.0",
-    "ethers": "6.10.0",
+    "ethers": "6.11.1",
```
Comment (clemsos) [inline governance/package.json:28]: Is that upgrade needed for the PR to work ? 

If not could you remove the change in version (better to have it in a separate PR as it may break other scripts on the repo).

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+    type: 'function',
+  },
+]
+
+/**
+ * Set up: instantiate L1 / L2 wallets connected to providers
+ */
+const walletPrivateKey = process.env.PRIVATE_KEY
+const L1RPC = 'https://rpc.unlock-protocol.com/1' // mainnet RPC
+const L2RPC = `https://arbitrum-mainnet.infura.io/v3/${process.env.ETHERSCAN_API_KEY}`
```
Comment (clemsos) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:121]: Here you can use the Unlock rpc too

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+      functionName: 'createRetryableTicket',
+      functionArgs: [
+        ARBTokenAddressOnL2, // to
+        0, // l2CallValue
+        L1ToL2MessageGasParams.maxSubmissionCost, // maxSubmissionCost
+        l2Wallet.address, // excessFeeRefundAddress
+        l2Wallet.address, // callValueRefundAddress
+        L1ToL2MessageGasParams.gasLimit, // gasLimit
+        gasPriceBid, // maxFeePerGas
+        transfer_calldata, // data
```
Comment (clemsos) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:236]: You can move these params to a variable as you are declaring them once before in `inbox_calldata`. Better to make sure both calls are consistent

File: governance/proposals/config.js [hunk 20]
```diff
@@ -0,0 +1,11 @@
+const ARBTokenAddressOnL2 = '0x912CE59144191C1204E64559FE8253a0e49E6548' // ARB TOKEN ADDRESS ON ARBITRUM ONE
+const grantsContractAddress = '0x00D5E0d31d37cc13C645D86410aB4cB7Cb428ccA'
+const timelockL2Alias = '0x28ffDfB0A6e6E06E95B3A1f928Dc4024240bD87c' // Timelock Alias Address on L2
+const L1TimelockContract = '0x17EEDFb0a6E6e06E95B3A1F928dc4024240BC76B' // Timelock Address mainnet
```
Comment (clemsos) [inline governance/proposals/config.js:4]: Could you capitalize constant names plz? It's a convention we try to follow across the repo

File: governance/.env.copy [hunk 21]
```diff
@@ -9,4 +9,4 @@ export BSCSCAN_API_KEY=
 export SNOWTRACE_API_KEY=
 export ARBISCAN_API_KEY=
 export CELO_API_KEY=
-export GNOSISSCAN_API_KEY=
\ No newline at end of file
+export GNOSISSCAN_API_KEY=
```
Comment (julien51) [inline governance/.env.copy:12]: Why is this in this PR?

File: governance/proposals/config.js [hunk 22]
```diff
@@ -0,0 +1,11 @@
+const ARB_TOKEN_ADRESS_ON_L2 = '0x912CE59144191C1204E64559FE8253a0e49E6548' // ARB TOKEN ADDRESS ON ARBITRUM ONE
```
Comment (julien51) [inline governance/proposals/config.js:1]: We already have that in the networks package, please is that!

File: governance/proposals/config.js [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+const GRANTS_CONTRACT_ADDRESS = '0x00D5E0d31d37cc13C645D86410aB4cB7Cb428ccA'
+const TIMELOCK_L2_ALIAS = '0x28ffDfB0A6e6E06E95B3A1f928Dc4024240bD87c' // Timelock Alias Address on L2
+const L1_TIMELOCK_CONTRACT = '0x17EEDFb0a6E6e06E95B3A1F928dc4024240BC76B' // Timelock Address mainnet
```
Comment (julien51) [inline governance/proposals/config.js:4]: Same here I am pretty sure it is already defined elsewhere. Please use that!

File: governance/proposals/config.js (hunk 22 above)
Comment (julien51) [inline governance/proposals/config.js:1]: this should also be named `constants.ts` instead of `config`

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+    stateMutability: 'payable',
+    type: 'function',
+  },
+]
+
+/**
+ * Set up: instantiate L1 / L2 wallets connected to providers
+ */
+const walletPrivateKey = process.env.PRIVATE_KEY
+const L1RPC = 'https://rpc.unlock-protocol.com/1' // mainnet RPC
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:120]: Please use the address loaded from the package network (it is this URL) but load it from the package network

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+const L2RPC = 'https://rpc.unlock-protocol.com/42161' // Arbitrum RPC
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:121]: same

**Review by julien51 — CHANGES_REQUESTED** [review]

File: governance/.env.copy (hunk 21 above)
Comment (blahkheart) [reply governance/.env.copy:12]: I didn't add that, it was already in the .env.copy file

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+const l1Provider = new ethers.JsonRpcProvider(L1RPC)
+const l2Provider = new ethers.JsonRpcProvider(L2RPC)
+// const l1Wallet = new ethers.Wallet(walletPrivateKey, l1Provider)
+const l2Wallet = new ethers.Wallet(walletPrivateKey, l2Provider)
+
+module.exports = async () => {
+  console.log('Cross-chain Proposer')
```
[the remark this answers is not in the document]
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:128]: I'm not sure what you mean here, what are you referring to please?

File: governance/proposals/config.js (hunk 22 above)
Comment (blahkheart) [reply governance/proposals/config.js:1]: Got it

File: governance/proposals/config.js (hunk 23 above)
Comment (blahkheart) [reply governance/proposals/config.js:4]: I'm not sure what exactly you are referring to here, please clarify?
For ```L1_TIMELOCK_CONTRACT```, I noticed it has been defined and used in several other files but I could not find it exported in any of them, to be used here.

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 27]
```diff
@@ -0,0 +1,239 @@
...
+      l2CallValue: 0,
+      excessFeeRefundAddress: L1_TIMELOCK_CONTRACT,
+      callValueRefundAddress: L1_TIMELOCK_CONTRACT,
+      data: transfer_calldata,
+    },
+    await getBaseFee(l1Provider),
+    l1Provider
+  )
+  const gasPriceBid = await l2Provider.getGasPrice()
+  const ETHDeposit = L1ToL2MessageGasParams.deposit.toNumber() * 10 // I Multiply by 10 to add extra in case gas changes due to proposal delay
```
[the remark this answers is not in the document]
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:176]: ETHDeposit is the amount ETH value that must be sent to the Delayed Inbox Contract for a successful retryable ticket creation. It is estimated by calling the ```l1ToL2MessageGasEstimate.estimateAll()``` function provided by the Arbitrum sdk. When the ```createRetryableTicket`` function is called, it forces the caller to pay this amount else reverts, this is why the ```excessFeeRefundAddress``` you asked about earlier is needed, to send any excess back. Here I multiply the actual value by 10 as a buffer because in the time it takes to execute a proposal that value may change.

File: governance/.gitignore [hunk 28]
```diff
@@ -2,6 +2,7 @@
 cache
 contracts
 artifacts
+.env
```
Comment (julien51) [inline governance/.gitignore:5]: same here

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 29]
```diff
@@ -0,0 +1,242 @@
...
+  const inboxAddress = ethBridger.l2Network.ethBridge.inbox
+
+  const L2TokenContract = new ethers.Contract(
+    ARB_TOKEN_ADRESS_ON_L2,
+    ERC20_ABI,
+    l2Wallet
+  ).connect(l2Wallet)
+
+  const balanceOf = await L2TokenContract.balanceOf(TIMELOCK_L2_ALIAS)
+  const tokenAmount = ethers.parseEther('1')
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:144]: Please let's use the *real* anount. It does not make sense to send 1 ARB if this has been tested correctly (as I am sure you have done by now!)

Also, please use `ethers.parseUnits` as this is an ERC20, not Ethers.

**Review by julien51 — CHANGES_REQUESTED** [review]

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 30]
```diff
@@ -0,0 +1,242 @@
...
+  )
+
+  const proposalName = `
+  # Test Transaction before 7k ARB Transfer To Fund Unlock Protocol’s Ecosystem via Grants Stack
+  
+  ### Goal of the proposal
+   This proposal requests to use 1 ARB from the tokens given to Unlock Protocol DAO by ArbitrumDAO to run a test transaction to de-risk the transfer of 7k ARB tokens to fund the retroQF round on Grants Stack.
+  
+  #### Current situation of DAO's ARB Tokens
+    - total: ${ethers.formatEther(balanceOf).toString()} ARB.
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:203]: same here. do not use `formatEther` when it is an ERC20.

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 31]
```diff
@@ -0,0 +1,242 @@
...
+    - DAO ALIAS Address (On Arbitrum): [${TIMELOCK_L2_ALIAS}](https://arbiscan.io/address/${TIMELOCK_L2_ALIAS})
+
+  For Reference
+  [Snapshot temperature check for 7k ARBs](https://snapshot.org/#/unlock-protocol.eth/proposal/0xaa142e599d981f0b58c3ac1a51af9f9a52fb5307f27d791ecc18c4da69eeacc3)
+  
+  #### About the proposal
+    The proposal contains a single call to the Arbitrum Delayed Inbox Contract's \`createRetryableTicket\` function on mainnet to create a \`Retryable Ticket\` that will attempt to execute an L2 request to the ARB token contract to transfer ${ethers
+      .formatEther(tokenAmount)
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:211]: same

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 32]
```diff
@@ -0,0 +1,242 @@
...
+      .toString()} of token from the Timelock L2 Alias address \`${TIMELOCK_L2_ALIAS}\` to the [grants contract](https://arbiscan.io/address/0x00d5e0d31d37cc13c645d86410ab4cb7cb428cca) - \`transfer(${GRANTS_CONTRACT_ADDRESS},${ethers
+      .formatEther(tokenAmount)
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:213]: same

**Review by julien51 — COMMENTED** [review]
since the DAO does not hold any ETH and yet it needs to send ETH to the mailbox contract (if I understand your proposal correctly), how will this work?

File: governance/.env.copy (hunk 21 above)
Comment (clemsos) [reply governance/.env.copy:12]: its a deleted newline it seems

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 33]
```diff
@@ -0,0 +1,242 @@
...
+      {
+        internalType: 'bool',
+        name: '',
+        type: 'bool',
+      },
+    ],
+    stateMutability: 'nonpayable',
+    type: 'function',
+  },
+]
```
Comment (clemsos) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:60]:
```suggestion
const ERC20_ABI = require('@unlock-protocol/hardhat-helpers/dist/ABIs/erc20.json')
```

File: governance/proposals/constants.js [hunk 34]
```diff
@@ -0,0 +1,19 @@
+const { mainnet, arbitrum } = require('@unlock-protocol/networks')
+
+const L1_RPC = mainnet.provider // mainnet RPC
+const L2_RPC = arbitrum.provider // Arbitrum RPC
+const GRANTS_CONTRACT_ADDRESS = '0x00D5E0d31d37cc13C645D86410aB4cB7Cb428ccA' // Grants contract on Arbitrum
+const TIMELOCK_L2_ALIAS = '0x28ffDfB0A6e6E06E95B3A1f928Dc4024240bD87c' // Timelock Alias Address on L2
+const L1_TIMELOCK_CONTRACT = '0x17EEDFb0a6E6e06E95B3A1F928dc4024240BC76B' // Timelock Address mainnet
+const ARB_TOKEN_ADRESS_ON_L2 = arbitrum.tokens.filter(
+  (token) => token.symbol === 'ARB'
+)[0].address // ARB TOKEN ADDRESS ON ARBITRUM ONE
```
Comment (clemsos) [inline governance/proposals/constants.js:10]:
```suggestion
// ARB TOKEN ADDRESS ON ARBITRUM ONE
const { address : ARB_TOKEN_ADRESS_ON_L2} = arbitrum.tokens.find(
  ({ symbol }) =>  === 'ARB'
)
```

File: governance/.env.copy (hunk 21 above)
Comment (julien51) [reply governance/.env.copy:12]: yeah that change should *not* be in this PR.

**Comment by clemsos** [comment]
@blahkheart I refactored the proposal script a bit to make it clearer. Almost there !

File: governance/package.json [hunk 35]
```diff
@@ -21,8 +23,10 @@
     "@unlock-protocol/hardhat-helpers": "workspace:^",
     "@unlock-protocol/hardhat-plugin": "workspace:^",
     "@unlock-protocol/networks": "workspace:./packages/networks",
+    "arb-shared-dependencies": "1.0.0",
     "eslint": "8.54.0",
     "ethers": "6.10.0",
+    "ethers5": "npm:ethers@5",
```
[the remark this answers is not in the document]
Comment (blahkheart) [reply governance/package.json:29]: I'm not sure how `"ethers5": "npm:ethers@5"` is in the package.json here, its not on mine and I don't think it's required to work. Is there any specific reason for you saying this?

**Comment by blahkheart** [comment]
> @blahkheart I refactored the proposal script a bit to make it clearer. Almost there !

Great! Is any further action required on my part?

**Comment by blahkheart** [comment]
> since the DAO does not hold any ETH and yet it needs to send ETH to the mailbox contract (if I understand your proposal correctly), how will this work?

I think I mentioned this earlier, one way I feel this could work is whoever calls the `execute` function on the governor has to provide the ETH to be sent. So, the DAO then funds/refunds the person, anyone can do this (I committed to do that on the test one, and I'm happy to do so again ) or the DAO can elect someone to do that and fund/refund them to do it.

File: governance/proposals/010-arbitrum-l1-l2-messaging.js [hunk 36]
```diff
@@ -0,0 +1,215 @@
...
+  const inboxCalldata = inboxContractInterface.encodeFunctionData(
+    'createRetryableTicket',
+    params
+  )
+
+  const proposalName = `
+  # Transfer 8200 ARB To Fund Unlock Protocol’s Ecosystem via Grants Stack
+  
+  ### Goal of the proposal
+   This proposal requests to use 8200 ARB from the tokens given to Unlock Protocol DAO by ArbitrumDAO to fund the retroQF round on Grants Stack for projects building on Unlock Protocol.
```
Comment (julien51) [inline governance/proposals/010-arbitrum-l1-l2-messaging.js:166]: This is UP from what the snapshot proposal initially asked...

> Asking for 7k in ARB to help Fund Unlock Protocol’s Ecosystem via Grants Stack

**Comment by julien51** [comment]
Please @blahkheart merge the conflicts on `yarn.lock`!

File: governance/proposals/010-arbitrum-l1-l2-messaging.js (hunk 36 above)
Comment (blahkheart) [reply governance/proposals/010-arbitrum-l1-l2-messaging.js:166]: There is an explanation for why this is up from the initial 7k as well as a reference to the snapshot proposal included in the description

**Review by julien51 — APPROVED** [review]
